lib/Doctrine/ORM/Tools/Console/Command/ClearCache/QueryCommand.php

```diff
@@ -63,32 +64,46 @@ protected function execute(InputInterface $input, OutputInterface $output)
     {
         $ui = (new SymfonyStyle($input, $output))->getErrorStyle();
 
-        $em          = $this->getEntityManager($input);
-        $cache       = $em->getConfiguration()->getQueryCache();
-        $cacheDriver = $em->getConfiguration()->getQueryCacheImpl();
+        $em    = $this->getEntityManager($input);
+        $cache = $em->getConfiguration()->getQueryCache();
 
-        if (! $cacheDriver) {
-            throw new InvalidArgumentException('No Query cache driver is configured on given EntityManager.');
-        }
-
-        if ($cacheDriver instanceof ApcCache || $cache instanceof ApcuAdapter) {
+        if ($cache instanceof ApcuAdapter) {
             throw new LogicException('Cannot clear APCu Cache from Console, it\'s shared in the Webserver memory and not accessible from the CLI.');
         }
 
-        if ($cacheDriver instanceof XcacheCache) {
-            throw new LogicException('Cannot clear XCache Cache from Console, it\'s shared in the Webserver memory and not accessible from the CLI.');
-        }
+        $cacheDriver = null;
+        if (! $cache) {
+            $cacheDriver = $em->getConfiguration()->getQueryCacheImpl();
+
+            if (! $cacheDriver) {
+                throw new InvalidArgumentException('No Query cache driver is configured on given EntityManager.');
+            }
+
+            if ($cacheDriver instanceof ApcCache) {
+                throw new LogicException('Cannot clear APCu Cache from Console, it\'s shared in the Webserver memory and not accessible from the CLI.');
+            }
 
-        if (! ($cacheDriver instanceof ClearableCache)) {
-            throw new LogicException(sprintf(
-                'Can only clear cache when ClearableCache interface is implemented, %s does not implement.',
-                get_debug_type($cacheDriver)
-            ));
+            if ($cacheDriver instanceof XcacheCache) {
+                throw new LogicException('Cannot clear XCache Cache from Console, it\'s shared in the Webserver memory and not accessible from the CLI.');
+            }
+
+            if (! ($cacheDriver instanceof ClearableCache)) {
+                throw new LogicException(sprintf(
+                    'Can only clear cache when ClearableCache interface is implemented, %s does not implement.',
+                    get_debug_type($cacheDriver)
+                ));
+            }
         }
 
         $ui->comment('Clearing <info>all</info> Query cache entries');
 
-        $result  = $cache ? $cache->clear() : $cacheDriver->deleteAll();
+        if ($cache) {
+            $result = $cache->clear();
+        } else {
+            assert($cacheDriver !== null);
```

Comment:
:thinking: is caching mandatory? Why couldn't both `$cache` and `$cacheDriver` be null?

Reply:
Previously there was 
```
$result = $cache ? $cache->clear() : $cacheDriver->deleteAll();
```

the fact is that with the refacto $cache and $cacheDriver cannot be null at the same time but unlike phpstan, psalm doesn't understand this so an assert is required.